libsolidity/CompilerStack.cpp

```diff
@@ -97,30 +97,41 @@ void CompilerStack::setSource(string const& _sourceCode)
 
 bool CompilerStack::parse()
 {
+	// todo not sure about clear. can contain warnings
 	m_errors.clear();
 
 	for (auto& sourcePair: m_sources)
 	{
 		sourcePair.second.scanner->reset();
-		sourcePair.second.ast = Parser().parse(sourcePair.second.scanner);
+		sourcePair.second.ast = Parser(m_errors).parse(sourcePair.second.scanner); // todo check for errors
 	}
+	if (!Error::containsOnlyWarnings(m_errors))
+		// errors while parsing. sould stop before type checking
+		return false;
+
 	resolveImports();
 
 	m_globalContext = make_shared<GlobalContext>();
-	NameAndTypeResolver resolver(m_globalContext->declarations());
+	bool success = true;
+	NameAndTypeResolver resolver(m_globalContext->declarations(), m_errors);
 	for (Source const* source: m_sourceOrder)
-		resolver.registerDeclarations(*source->ast);
+		success = success && resolver.registerDeclarations(*source->ast);
```

Comment:
What about `if (!resolver.registerDeclarations(*source->ast)) return false;`?
I think the current way is a bit misleading becouse of short-circuiting.